### PR TITLE
Add stream compression utility

### PR DESCRIPTION
**Pull request by sandorw**
No docs update at the moment - happy to add something to the release notes or wait for the compressed stream store implementation that will make use of this.

---
This change is [Reviewable](https://reviewable.io/reviews/palantir/atlasdb/1312)

### CODE REVIEW

**Comment by sandorw** [comment]
@clockfort @schlosna @nziebart to take a look

File: atlasdb-commons/src/main/java/com/palantir/common/compression/LZ4DecompressingInputStream.java [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+    @Override
+    public void close() throws IOException {
+        delegate.close();
+    }
+
+    // This reads one block plus the size of the next block in as few
+    // read calls as possible. The length should be the number of bytes
+    // of data plus HEADER_SIZE bytes for the length of the next block.
+    private int readBlock(byte[] b, int length) throws IOException {
+        int read = 0;
```
Comment (gsheasby) [inline atlasdb-commons/src/main/java/com/palantir/common/compression/LZ4DecompressingInputStream.java]: I'd prefer `readSoFar`, as it's difficult to infer tense from a single-word variable.

File: atlasdb-commons/src/main/java/com/palantir/common/compression/LZ4CompressingInputStream.java [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+            int remainingBuffer = maxPosition - position;
+            int bytesToRead = Math.min(len - bytesRead, remainingBuffer);
+            System.arraycopy(compressedBuffer, position, b, off + bytesRead, bytesToRead);
+            position += bytesToRead;
+            bytesRead += bytesToRead;
+            if ((position == maxPosition) && !refill()) {
+                break;
+            }
+        }
+        return bytesRead > 0 ? bytesRead : -1;
```
Comment (gsheasby) [inline atlasdb-commons/src/main/java/com/palantir/common/compression/LZ4CompressingInputStream.java]: General style nit: leave a blank line after a long (>1 line) if/while/for/etc block.

File: atlasdb-commons/src/main/java/com/palantir/common/compression/LZ4DecompressingInputStream.java [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+    }
+
+    @Override
+    public void close() throws IOException {
+        delegate.close();
+    }
+
+    // This reads one block plus the size of the next block in as few
+    // read calls as possible. The length should be the number of bytes
+    // of data plus HEADER_SIZE bytes for the length of the next block.
```
Comment (gsheasby) [inline atlasdb-commons/src/main/java/com/palantir/common/compression/LZ4DecompressingInputStream.java]: Good comment, this is the kind of description I'd like to see elsewhere.

File: atlasdb-commons/src/main/java/com/palantir/common/compression/LZ4DecompressingInputStream.java [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+            return false;
+        }
+
+        if (length > 0) {
+            //Compressed block
+            nextLength = readBlock(compressedBuffer, length + LZ4Streams.HEADER_SIZE);
+            maxPosition = decompressor.decompress(compressedBuffer, 0, length, buffer, 0, frameDescriptor.maximumBlockSize);
+        } else {
+            //Uncompressed block
+            length ^= 0x80000000;
```
Comment (gsheasby) [inline atlasdb-commons/src/main/java/com/palantir/common/compression/LZ4DecompressingInputStream.java]: this line could use explanation.

**Comment by sandorw** [comment]
I made some extensive refactors and took your style recommendations as well as adding more comments. There's now a common abstract stream class that includes the read logic and the (de)compressing streams define the logic for refilling the internal buffer.

The test refactor isn't finished yet - most of the tests for the compressing stream are written, but I want to write similar tests for the decompressing stream as well as some tests using both together. I'll finish those tomorrow, but any feedback in the meantime is welcome. Thoughts on overall test strategy are also welcome. Perhaps it would be best to split the longer tests up into smaller ones to make it clear what's being tested and therefore what's failing.

File: atlasdb-commons/src/main/java/com/palantir/common/compression/LZ4CompressingInputStream.java [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+        return true;
+    }
+
+    private void writeFooter() throws IOException {
+        // Write 0 for the size of the next block to indicate no blocks are left
+        writeBlockSize(0);
+        position = 0;
+        maxPosition = LZ4Streams.BLOCK_HEADER_LENGTH;
+        if (frameDescriptor.hasContentChecksum) {
+            // Write the final content hash
```
Comment (gsheasby) [inline atlasdb-commons/src/main/java/com/palantir/common/compression/LZ4CompressingInputStream.java]: Maybe you could extract the 3 lines to `writeFinalContentHash`?

File: atlasdb-commons/src/main/java/com/palantir/common/compression/LZ4CompressingInputStream.java [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+            System.arraycopy(hashValue, 0, buffer, LZ4Streams.BLOCK_HEADER_LENGTH, 4);
+            maxPosition = LZ4Streams.BLOCK_HEADER_LENGTH + 4;
+        }
+    }
+
+    private void writeBlockSize(int length) throws IOException {
+        byte[] blockSize = LZ4Streams.intToLittleEndianBytes(length);
+        copyToStartOfBuffer(blockSize, 0, 4);
+    }
+
```
Comment (gsheasby) [inline atlasdb-commons/src/main/java/com/palantir/common/compression/LZ4CompressingInputStream.java]: nit: double blank line here

File: atlasdb-commons/src/main/java/com/palantir/common/compression/LZ4DecompressingInputStream.java [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+                LZ4Streams.MAGIC_LENGTH, LZ4Streams.MAGIC_LENGTH + LZ4Streams.FRAME_DESCRIPTOR_LENGTH));
+    }
+
+    // This reads one block plus the size of the next block in as few
+    // read calls as possible. The length should be the number of bytes
+    // of data plus HEADER_SIZE bytes for the length of the next block.
+    private int readBlock(byte[] b, int length) throws IOException {
+        int bytesReadSoFar = 0;
+        while (bytesReadSoFar < length) {
+            int numRead = delegate.read(b, bytesReadSoFar, length - bytesReadSoFar);
```
Comment (gsheasby) [inline atlasdb-commons/src/main/java/com/palantir/common/compression/LZ4DecompressingInputStream.java]: Is there a case where you ask `delegate` to give you `N` bytes, it gives you `M` bytes (where 0 < `M` < `N`), and then the next time around, it still gives you some bytes? Sounds a bit odd to me, but maybe I'm missing something here.

File: atlasdb-commons/src/main/java/com/palantir/common/compression/LZ4FrameDescriptor.java [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+    private static final int[] BLOCK_SIZE_LOOKUP = new int[] {
+            -1, -1, -1, -1, // 0-3 are N/A
+            1 << 16, // 64 KB
+            1 << 18, // 256 KB
+            1 << 20, // 1 MB
+            1 << 22  // 4 MB
+    };
+
+    private final int maximumBlockSizeIndex;
+    final int maximumBlockSize;
```
Comment (gsheasby) [inline atlasdb-commons/src/main/java/com/palantir/common/compression/LZ4FrameDescriptor.java]: Why aren't these two private?

File: atlasdb-commons/src/main/java/com/palantir/common/compression/LZ4FrameDescriptor.java [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+    public static LZ4FrameDescriptor fromByteArray(byte[] descriptor) {
+        boolean hasContentChecksum = (descriptor[0] & CONTENT_CHECKSUM_FLAG) > 0;
+        int blockSize = (descriptor[1] & 0x70) >> 4;
+        return new LZ4FrameDescriptor(hasContentChecksum, blockSize);
+    }
+
+    public LZ4FrameDescriptor(boolean hasContentChecksum, int maximumBlockSizeIndex) {
+        this.hasContentChecksum = hasContentChecksum;
+        this.maximumBlockSizeIndex = maximumBlockSizeIndex;
+        Preconditions.checkArgument(maximumBlockSizeIndex > 3 && maximumBlockSizeIndex < 8,
```
Comment (gsheasby) [inline atlasdb-commons/src/main/java/com/palantir/common/compression/LZ4FrameDescriptor.java]: If possible, Preconditions checks should be done before we set fields.

File: atlasdb-commons/src/main/java/com/palantir/common/compression/LZ4FrameDescriptor.java [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+        LZ4FrameDescriptor that = (LZ4FrameDescriptor) o;
+        if (hasContentChecksum != that.hasContentChecksum) {
+            return false;
+        }
+        return maximumBlockSizeIndex == that.maximumBlockSizeIndex;
+    }
+
+    @Override
+    public int hashCode() {
+        int result = (hasContentChecksum ? 1 : 0);
```
Comment (gsheasby) [inline atlasdb-commons/src/main/java/com/palantir/common/compression/LZ4FrameDescriptor.java]: brackets are unnecessary here.

File: atlasdb-commons/src/main/java/com/palantir/common/compression/BufferedDelegateInputStream.java [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+    public final int read(byte b[], int off, int len) throws IOException {
+        if (b == null) {
+            throw new NullPointerException();
+        } else if (off < 0 || len < 0 || len > b.length - off) {
+            throw new IndexOutOfBoundsException();
+        } else if (len == 0) {
+            return 0;
+        }
+
+        // Read from the buffer until the delegate stream is exhausted
```
Comment (sandorw) [inline atlasdb-commons/src/main/java/com/palantir/common/compression/BufferedDelegateInputStream.java]: This looks good with the caveat that I want to update the position as close to any System.arraycopy calls as possible.

File: atlasdb-commons/src/main/java/com/palantir/common/compression/LZ4DecompressingInputStream.java (hunk 7 above)
Comment (sandorw) [inline atlasdb-commons/src/main/java/com/palantir/common/compression/LZ4DecompressingInputStream.java]: I think the docs on InputStream suggest that this shouldn't happen. https://docs.oracle.com/javase/7/docs/api/java/io/InputStream.html#read(byte[],%20int,%20int) indicates that the read call should block until bytes are available. The default, inefficient, implementation blocks until the requested number of bytes is read or end of files is detected.

Alternatively, we could read from the delegate until a -1 is explicitly received, but I don't think this is necessary.

File: atlasdb-commons/src/main/java/com/palantir/common/compression/LZ4FrameDescriptor.java (hunk 8 above)
Comment (sandorw) [inline atlasdb-commons/src/main/java/com/palantir/common/compression/LZ4FrameDescriptor.java]: I'll make them private and add getters.

**Comment by nziebart** [comment]
Really high level question: is there a reason we can't use
LZ4BlockInputStream and LZ4BlockOutputStream?

**Comment by sandorw** [comment]
The atlas stream interface acts on InputStreams both on the reading and writing side, so we need a compressing stream that takes and produces an InputStream rather than an OutputStream. The LZ4BlockInputStream should theoretically work with my compressed streams if they both conform to the spec, but I haven't tried them - that implementation is certainly more complicated and I haven't puzzled out the compression levels and methods mean. Perhaps we can take a look at it again tomorrow.

Tests should be complete. Aside from addressing CR comments, the non-test classes are largely unchanged. One behavior change I made was for the decompressing stream to perform the checksum validation as soon as the last bytes are read, rather than requiring one to attempt to read from an empty stream or over-read to validate the contents.

**Comment by gsheasby** [comment]
Reviewed 1 of 9 files at r1, 4 of 6 files at r3, 2 of 3 files at r4.
Review status: 7 of 9 files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[atlasdb-commons/src/main/java/com/palantir/common/compression/BufferedDelegateInputStream.java, line 86 at r4](https://reviewable.io:443/reviews/palantir/atlasdb/1312#-KY-QvpswH7rUROljy_5:-KY-QvptCSdcMymOjx7w:b-6rxx45) ([raw file](https://github.com/palantir/atlasdb/blob/a6f271bf94babef3d3390fbec7fb98501f33d827/atlasdb-commons/src/main/java/com/palantir/common/compression/BufferedDelegateInputStream.java#L86)):*
> ```Java
>         while (bytesReadSoFar < len) {
>             int bytesRead = readBytes(b, off, len, bytesReadSoFar);
>             bytesReadSoFar += bytesRead;
> ```

Now you can inline this :-)

---

*[atlasdb-commons/src/main/java/com/palantir/common/compression/LZ4FrameDescriptor.java, line 51 at r4](https://reviewable.io:443/reviews/palantir/atlasdb/1312#-KY-TIQtN_G0QDA86sN5:-KY-TIQtN_G0QDA86sN6:b-v2va0j) ([raw file](https://github.com/palantir/atlasdb/blob/a6f271bf94babef3d3390fbec7fb98501f33d827/atlasdb-commons/src/main/java/com/palantir/common/compression/LZ4FrameDescriptor.java#L51)):*
> ```Java
>     private static final byte START_BD_FLAG_BYTE = 0x00;
>     private static final byte CONTENT_CHECKSUM_FLAG = 1 << 2;
>     private static final int[] BLOCK_SIZE_LOOKUP = new int[] {
> ```

When I started looking through the tests, I was wondering what 5 was. That lead me here.

This is..... opaque. How about an enum `BlockSize`, so you can say something like `new LZ4FrameDescriptor(true, BlockSize.256_KB);`, rather than `new LZ4FrameDescriptor(true, 5);`?

Note that `BlockSize.256_KB` won't work as you'll need to start with a letter. `SIZE_256_KB`? `KB_256`? `TWO_FIFTY_SIX_KB`? IMHO, any of those says "256KB" better than `5` does.

---

*[atlasdb-commons/src/test/java/com/palantir/common/compression/LZ4CompressingInputStreamTests.java, line 42 at r4](https://reviewable.io:443/reviews/palantir/atlasdb/1312#-KY-URG80Q_EtCr3Zg0z:-KY-URG9BwzGQWu9TKYM:b-efkh1m) ([raw file](https://github.com/palantir/atlasdb/blob/a6f271bf94babef3d3390fbec7fb98501f33d827/atlasdb-commons/src/test/java/com/palantir/common/compression/LZ4CompressingInputStreamTests.java#L42)):*
> ```Java
>     private static final byte[] INCOMPRESSIBLE_MULTIBLOCK_DATA = new byte[512 * 1024];
>     private static final byte[] MIX_MULTIBLOCK_DATA = new byte[512 * 1024];
>     private static final LZ4FrameDescriptor CHECKSUM_FRAME_DESCRIPTOR = new LZ4FrameDescriptor(true, 5);
> ```

5? (see other comment on block sizes)

---

*[atlasdb-commons/src/test/java/com/palantir/common/compression/LZ4CompressingInputStreamTests.java, line 47 at r4](https://reviewable.io:443/reviews/palantir/atlasdb/1312#-KY-VTYIPl_yaAK0rQmq:-KY-VTYIPl_yaAK0rQmr:bj0cvy7) ([raw file](https://github.com/palantir/atlasdb/blob/a6f271bf94babef3d3390fbec7fb98501f33d827/atlasdb-commons/src/test/java/com/palantir/common/compression/LZ4CompressingInputStreamTests.java#L47)):*
> ```Java
>         Arrays.fill(COMPRESSIBLE_MULTIBLOCK_DATA, (byte) 0xFF);
>         new Random(0).nextBytes(INCOMPRESSIBLE_MULTIBLOCK_DATA);
>         System.arraycopy(COMPRESSIBLE_MULTIBLOCK_DATA, 0, MIX_MULTIBLOCK_DATA, 0, 256 * 1024);
> ```

It's worth extracting `BLOCK_SIZE = 256 * 1024`, and then using `2 * BLOCK_SIZE` instead of `512 * 1024` above.

---

*[atlasdb-commons/src/test/java/com/palantir/common/compression/LZ4CompressingInputStreamTests.java, line 66 at r4](https://reviewable.io:443/reviews/palantir/atlasdb/1312#-KY-Vwqp2KU2B5ISDWYC:-KY-Vwqp2KU2B5ISDWYD:b38oeru) ([raw file](https://github.com/palantir/atlasdb/blob/a6f271bf94babef3d3390fbec7fb98501f33d827/atlasdb-commons/src/test/java/com/palantir/common/compression/LZ4CompressingInputStreamTests.java#L66)):*
> ```Java
> 
>         byte[] data = new byte[100];
>         int bytesRead = compressedStream.read(data, 0, 0);
> ```

Shouldn't we also try to read more than zero bytes?

---

*[atlasdb-commons/src/test/java/com/palantir/common/compression/LZ4CompressingInputStreamTests.java, line 72 at r4](https://reviewable.io:443/reviews/palantir/atlasdb/1312#-KY-Wo2RCq1YqtkfjoKD:-KY-Wo2RCq1YqtkfjoKE:b-30s3tl) ([raw file](https://github.com/palantir/atlasdb/blob/a6f271bf94babef3d3390fbec7fb98501f33d827/atlasdb-commons/src/test/java/com/palantir/common/compression/LZ4CompressingInputStreamTests.java#L72)):*
> ```Java
> 
>     @Test
>     public void testFrameHeaderFormat_defaultFrameDescriptor() throws IOException {
> ```

This and the below test are almost identical - let's extract a method with the common code.

---

*[atlasdb-commons/src/test/java/com/palantir/common/compression/LZ4CompressingInputStreamTests.java, line 204 at r4](https://reviewable.io:443/reviews/palantir/atlasdb/1312#-KY-ZHOUohDYqmKw9yzR:-KY-ZHOVMvXBlW4sQmXW:b-19ojeu) ([raw file](https://github.com/palantir/atlasdb/blob/a6f271bf94babef3d3390fbec7fb98501f33d827/atlasdb-commons/src/test/java/com/palantir/common/compression/LZ4CompressingInputStreamTests.java#L204)):*
> ```Java
>         }
> 
>         int finalBlockSize = readNextBlockSize();
> ```

These three lines repeat a few times. Another helper method?

---

*[atlasdb-commons/src/test/java/com/palantir/common/compression/LZ4CompressingInputStreamTests.java, line 251 at r4](https://reviewable.io:443/reviews/palantir/atlasdb/1312#-KY-ZbodZY5ob2FnQhAP:-KY-ZbodZY5ob2FnQhAQ:b-lesvj5) ([raw file](https://github.com/palantir/atlasdb/blob/a6f271bf94babef3d3390fbec7fb98501f33d827/atlasdb-commons/src/test/java/com/palantir/common/compression/LZ4CompressingInputStreamTests.java#L251)):*
> ```Java
>         // Compressed blocks
>         for (int i = 0; i < numBlocks / 2; ++i) {
>             int blockSize = readNextBlockSize();
> ```

Another copy/pasted for loop. Let's extract a method, taking in number of repeats, `expectedCompressedBlockSize`, and `expectedCompressedBlock`.

---

*[atlasdb-commons/src/test/java/com/palantir/common/compression/LZ4CompressingInputStreamTests.java, line 259 at r4](https://reviewable.io:443/reviews/palantir/atlasdb/1312#-KY-Zol0BfhGqPFnxV0y:-KY-Zol0BfhGqPFnxV0z:b-7vrt8x) ([raw file](https://github.com/palantir/atlasdb/blob/a6f271bf94babef3d3390fbec7fb98501f33d827/atlasdb-commons/src/test/java/com/palantir/common/compression/LZ4CompressingInputStreamTests.java#L259)):*
> ```Java
>         // Uncompressed blocks
>         for (int i = 0; i < numBlocks / 2; ++i) {
>             int blockSize = readNextBlockSize();
> ```

Same as above - this for loop has a very close friend :-)

---

*[atlasdb-commons/src/test/java/com/palantir/common/compression/LZ4CompressingInputStreamTests.java, line 332 at r4](https://reviewable.io:443/reviews/palantir/atlasdb/1312#-KY-_Vu9EjVvyzly8BAQ:-KY-_Vu9EjVvyzly8BAR:b-lw2y4d) ([raw file](https://github.com/palantir/atlasdb/blob/a6f271bf94babef3d3390fbec7fb98501f33d827/atlasdb-commons/src/test/java/com/palantir/common/compression/LZ4CompressingInputStreamTests.java#L332)):*
> ```Java
>     // Terminates after a zero size block size is read.
>     private void skipDataBlocks() throws IOException {
>         int blockSize = readNextBlockSize();
> ```

You can save a line (and avoid repetition) with a `do { ... } while ();` loop.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/atlasdb/1312)*

**Comment by gsheasby** [comment]
Reviewed 1 of 11 files at r2, 1 of 3 files at r4.
Review status: all files reviewed at latest revision, 16 unresolved discussions, some commit checks failed.

---

*[atlasdb-commons/src/test/java/com/palantir/common/compression/LZ4DecompressingInputStreamTests.java, line 53 at r4](https://reviewable.io:443/reviews/palantir/atlasdb/1312#-KY-m7FoNsZ0weeyNQxT:-KY-m7FpjqSMRKawYZmY:b-i961cv) ([raw file](https://github.com/palantir/atlasdb/blob/a6f271bf94babef3d3390fbec7fb98501f33d827/atlasdb-commons/src/test/java/com/palantir/common/compression/LZ4DecompressingInputStreamTests.java#L53)):*
> ```Java
>     @Before
>     public void before() throws IOException {
>         compressedBuffer = new byte[512 * 1024];
> ```

magic number

---

*[atlasdb-commons/src/test/java/com/palantir/common/compression/LZ4DecompressingInputStreamTests.java, line 111 at r4](https://reviewable.io:443/reviews/palantir/atlasdb/1312#-KY-nRuU9qyinfJHKMLT:-KY-nRuU9qyinfJHKMLU:b46vpng) ([raw file](https://github.com/palantir/atlasdb/blob/a6f271bf94babef3d3390fbec7fb98501f33d827/atlasdb-commons/src/test/java/com/palantir/common/compression/LZ4DecompressingInputStreamTests.java#L111)):*
> ```Java
>         writeFrameHeader(LZ4Streams.DEFAULT_FRAME_DESCRIPTOR);
>         writeBlockSize(toUncompressedBlockSize(2));
>         byte firstValue = (byte) 42;
> ```

These are repeated a bunch (including in the `testSingleByte` test). Constants `FIRST_BYTE` and `SECOND_BYTE`?

---

*[atlasdb-commons/src/test/java/com/palantir/common/compression/LZ4DecompressingInputStreamTests.java, line 200 at r4](https://reviewable.io:443/reviews/palantir/atlasdb/1312#-KY-pEGixo8ssbY2fxIl:-KY-pEGjn2B4mVZDeb-B:bgsewt4) ([raw file](https://github.com/palantir/atlasdb/blob/a6f271bf94babef3d3390fbec7fb98501f33d827/atlasdb-commons/src/test/java/com/palantir/common/compression/LZ4DecompressingInputStreamTests.java#L200)):*
> ```Java
>         writeBlockSize(COMPRESSED_DATA_B.length);
>         writeBlock(COMPRESSED_DATA_B);
>         writeBlockSize(0);
> ```

What happens if I forget to write an empty block size before initialising the stream?

---

*[atlasdb-commons/src/test/java/com/palantir/common/compression/LZ4DecompressingInputStreamTests.java, line 248 at r4](https://reviewable.io:443/reviews/palantir/atlasdb/1312#-KY-pbFt6u0_QrGMsf8z:-KY-pbFt6u0_QrGMsf9-:b-7endt2) ([raw file](https://github.com/palantir/atlasdb/blob/a6f271bf94babef3d3390fbec7fb98501f33d827/atlasdb-commons/src/test/java/com/palantir/common/compression/LZ4DecompressingInputStreamTests.java#L248)):*
> ```Java
>         int expectedHashValue = hash.getValue();
> 
>         writeFrameHeader(CHECKSUM_FRAME_DESCRIPTOR);
> ```

This block is copied below, albeit with a different hash value. Another method to extract?

---

*[atlasdb-commons/src/test/java/com/palantir/common/compression/LZ4DecompressingInputStreamTests.java, line 285 at r4](https://reviewable.io:443/reviews/palantir/atlasdb/1312#-KY-pjFS-skaq1dpeli8:-KY-pjFS-skaq1dpeli9:b-amittk) ([raw file](https://github.com/palantir/atlasdb/blob/a6f271bf94babef3d3390fbec7fb98501f33d827/atlasdb-commons/src/test/java/com/palantir/common/compression/LZ4DecompressingInputStreamTests.java#L285)):*
> ```Java
>     @Test
>     public void testMultiBlock_compressedByCompressingStream_withChecksum() throws IOException {
>         byte[] uncompressedData = new byte[512 * 1024];
> ```

magic number (also below)

---

*[atlasdb-commons/src/test/java/com/palantir/common/compression/LZ4FrameDescriptorTests.java, line 26 at r4](https://reviewable.io:443/reviews/palantir/atlasdb/1312#-KY-lpVy5NiZ_kRo910Q:-KY-lpVy5NiZ_kRo910R:b8slsme) ([raw file](https://github.com/palantir/atlasdb/blob/a6f271bf94babef3d3390fbec7fb98501f33d827/atlasdb-commons/src/test/java/com/palantir/common/compression/LZ4FrameDescriptorTests.java#L26)):*
> ```Java
> 
>     @Test
>     public void testSerialization() {
> ```

nit: test could be split into two for easier debugging.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/atlasdb/1312)*

**Comment by nziebart** [comment]
Spoke with @sandorw a bit offline. It's probably worth at least considering if we can use `LZ4BlockOutputStream` here, since it's an existing part of the lz4 library. The issue is that we'd need to pipe the given `InputStream` to the `OutputStream` and pipe that back into an `InputStream` (i.e., have some mechanism where a read from the end `InputStream` triggers a read from the input `InputStream` that is then transformed through `LZ4BlockOutputStream`. 

@clockfort @schlosna for SA

File: atlasdb-commons/src/main/java/com/palantir/common/compression/LZ4DecompressingInputStream.java (hunk 7 above)
Comment (sandorw) [inline atlasdb-commons/src/main/java/com/palantir/common/compression/LZ4DecompressingInputStream.java]: Actually, it looks like I was mistaken here. The docs are still not very clear on the expected behavior, but I've found InputStream implementations that do not attempt to fully fulfill incoming read requests despite not hitting the end of the stream. For example, the LZ4 library LZ4BlockInputStream does this. It looks like the correct solution is to use ByteStreams.read wherever this behavior is expected. Notably, AbstractPersistentStreamStore already uses this when extracting InputStream bytes into blocks in the DB.

**Comment by sandorw** [comment]
Closed in favor of https://github.com/palantir/atlasdb/pull/1347

**Pull request closed without merging**